### PR TITLE
Spring core

### PR DESCRIPTION
**Pull request by shklyarskyvolodymyr**

### CODE REVIEW

File: src/main/java/movie/app/dao/impl/MovieListDAOImp.java [hunk 1]
```diff
@@ -39,4 +43,8 @@ public MovieList getUserMovie(int userId, int movieId) {
     query.setParameter("userId", userId);
     return (MovieList) query.getSingleResult();
   }
+
```
Comment (Nazariii) [inline src/main/java/movie/app/dao/impl/MovieListDAOImp.java]: We do not need `remove` method as it is in AbstractDao

File: src/test/java/movie/app/service/impl/UserServiceImpTest.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import movie.app.config.SpringConfig;
+import movie.app.model.User;
+import movie.app.service.UserService;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = SpringConfig.class)
```
Comment (Nazariii) [inline src/test/java/movie/app/service/impl/UserServiceImpTest.java]: Test is failing with `Table 'db.user' doesn't exist` during clean run. We need to setup tests metadata and data in in-memory DB for running it.

File: src/main/java/movie/app/config/SpringConfig.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package movie.app.config;
+
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ComponentScan("movie")
```
Comment (Nazariii) [inline src/main/java/movie/app/config/SpringConfig.java]: please update you package structure according to https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html eg: `com.movieapp`

File: src/test/java/org/movie/app/service/impl/UserServiceImpTest.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+@ContextConfiguration(classes = SpringConfig.class)
+class UserServiceImpTest {
+
+    private String login = "Vasya555";
+    private int subscriptionId = 1;
+    @Autowired
+    private UserService userService;
+
+    // Testing if UserService.getUserByLogin correctly return user by login
+    @Test
```
Comment (Nazariii) [inline src/test/java/org/movie/app/service/impl/UserServiceImpTest.java]: For unit testing please use these best practices: https://kb.epam.com/display/EJAVACC/Unit+Testing+Best+Practices

File: src/test/java/org/movie/app/service/impl/SubscriptionServiceImpTest.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = SpringConfig.class)
+class SubscriptionServiceImpTest {
```
Comment (Nazariii) [inline src/test/java/org/movie/app/service/impl/SubscriptionServiceImpTest.java]: Please, separate integration test into a separate folder: `src/it` and run them by `failsafe` plugin. https://maven.apache.org/surefire/maven-failsafe-plugin/usage.html

File: src/main/java/org/movie/app/PersistanceUtil.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.stereotype.Component;
+
+import javax.persistence.EntityManager;
+
+/**
+ * Class that maintain Session object
+ */
+@Component
+@RequiredArgsConstructor
+public class PersistanceUtil {
```
Comment (Nazariii) [inline src/main/java/org/movie/app/PersistanceUtil.java]: Please, remove PersistanceUtil class and use `@PersistenceContext` annotation to autowire enitityManager.
```
@PersistenceContext
private EntityManager persistence;
```

File: src/main/java/org/movie/app/service/impl/MovieListServiceImp.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     * Add movie into user list by MovieId
+     */
+    public void addMovieToUserList(UserDTO userDTO, int movieId) {
+        MovieList movieList = new MovieList();
+        try {
+            persistence.getEntityManager().getTransaction().begin();
+            movieList.setMovieListPK(new MovieListPK(userRepository.findByLogin(userDTO.getLogin()),
+                    movieRepository.findById(movieId)));
+            persistence.getEntityManager().persist(movieList);
```
Comment (Nazariii) [inline src/main/java/org/movie/app/service/impl/MovieListServiceImp.java]: We have missed one very important concept of DAO pattern. In general, best practice is to not expose Data Access logic to service layer (you mixing layers) so we need to rewrite logic so EntityManager lives only on DAO layer.
All DAO interfaces should be in the same hierarchy. Eg:
```
GenericDAO -> MovieDAO, LanguageDAO, ProducerDAO etc.
```
so for eg MovieDAO declaration should be:
```
public interface MovieDAO extends GenericDAO<Movie, Integer>
```
so we can rewrite `persistence.getEntityManager().persist(movieList);` to `movieListRepository.persist(movieList);`

File: src/main/java/org/movie/app/service/impl/MovieListServiceImp.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            persistence.getEntityManager().getTransaction().commit();
+        } catch (Exception e) {
+            persistence.getEntityManager().getTransaction().rollback();
```
Comment (Nazariii) [inline src/main/java/org/movie/app/service/impl/MovieListServiceImp.java]: Also, we can use Spring @org.springframework.transaction.annotation.Transactional annotation for transaction management and remove manual handling. But we need to configure `transactionManager`

File: src/main/java/org/movie/app/service/impl/MovieListServiceImp.java [hunk 9]
```diff
@@ -1,67 +1,115 @@
...
-      Movie movie = movieService.getMovie(movieId);
-      movieList.setMovieListPK(new MovieListPK(user, movie));
-      PersistanceUtil.getSession().persist(movieList);
-      PersistanceUtil.getSession().getTransaction().commit();
-    } catch (Exception e) {
-      PersistanceUtil.getSession().getTransaction().rollback();
+    /**
+     * update last view location in MovieList object
+     */
+    public void saveViewLocation(MovieListDTO movieListDTO, int time) {
```
[the remark this answers is not in the document]
Comment (shklyarskyvolodymyr) [reply src/main/java/org/movie/app/service/impl/MovieListServiceImp.java]: Done

File: src/main/java/org/movie/app/service/impl/MovieListServiceImp.java (hunk 7 above)
Comment (shklyarskyvolodymyr) [reply src/main/java/org/movie/app/service/impl/MovieListServiceImp.java]: Done

File: src/main/java/org/movie/app/PersistanceUtil.java (hunk 6 above)
Comment (shklyarskyvolodymyr) [reply src/main/java/org/movie/app/PersistanceUtil.java]: Done

File: src/main/java/org/movie/app/service/impl/MovieListServiceImp.java (hunk 8 above)
Comment (shklyarskyvolodymyr) [reply src/main/java/org/movie/app/service/impl/MovieListServiceImp.java]: Done

File: src/main/java/org/movie/app/service/impl/MovieListServiceImp.java [hunk 10]
```diff
@@ -1,67 +1,115 @@
...
-    try {
-      PersistanceUtil.getSession().getTransaction().begin();
-      MovieList mv = movieListRepository.getUserMovie(user.getUserId(), movie);
-      mv.setRating(rating);
-      movieListRepository.persist(mv);
-      PersistanceUtil.getSession().getTransaction().commit();
-    } catch (Exception e) {
-      PersistanceUtil.getSession().getTransaction().rollback();
+
+    private MovieListDTO convertToDTO(MovieList movieList) {
```
[the remark this answers is not in the document]
Comment (shklyarskyvolodymyr) [reply src/main/java/org/movie/app/service/impl/MovieListServiceImp.java]: Done

File: src/test/resources/META-INF/persistence.xml [hunk 11]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (shklyarskyvolodymyr) [reply src/test/resources/META-INF/persistence.xml]: done

File: src/test/java/org/movie/app/service/impl/UserServiceImpTest.java (hunk 4 above)
Comment (shklyarskyvolodymyr) [reply src/test/java/org/movie/app/service/impl/UserServiceImpTest.java]: done

File: src/test/java/org/movie/app/service/impl/SubscriptionServiceImpTest.java (hunk 5 above)
Comment (shklyarskyvolodymyr) [reply src/test/java/org/movie/app/service/impl/SubscriptionServiceImpTest.java]: Done

**Pull request merged**